File: backend/app/schemas/curriculum.py
# backend/app/schemas/curriculum.py
"""
Curriculum Pydantic Schemas

These schemas define the curriculum structure used across the application.
They match the frontend TypeScript types exactly to ensure end-to-end type safety.
"""

from pydantic import BaseModel, Field
from typing import Optional


# ============================================================================
# CORE CURRICULUM SCHEMAS
# ============================================================================

class CurriculumUnit(BaseModel):
    """Curriculum unit schema"""
    id: str = Field(..., description="Unit ID (e.g., 'SS001', 'COUNT001')")
    title: str = Field(..., description="Unit title")
    description: Optional[str] = Field(None, description="Unit description")

    class Config:
        json_schema_extra = {
            "example": {
                "id": "SS001",
                "title": "Classroom Routines and Social Skills",
                "description": "Classroom Routines and Social Skills"
            }
        }


class CurriculumSkill(BaseModel):
    """Curriculum skill schema"""
    id: str = Field(..., description="Skill ID (e.g., 'SS001-04', 'COUNT001-01')")
    description: str = Field(..., description="Skill description")

    class Config:
        json_schema_extra = {
            "example": {
                "id": "SS001-04",
                "description": "Local Governance"
            }
        }


class CurriculumSubskill(BaseModel):
    """Curriculum subskill schema"""
    id: str = Field(..., description="Subskill ID (e.g., 'SS001-04-E', 'COUNT001-01-A')")
    description: str = Field(..., description="Subskill description")

    class Config:
        json_schema_extra = {
            "example": {
                "id": "SS001-04-E",
                "description": "Create simple maps showing locations of local public services"
            }
        }


class CurriculumMetadata(BaseModel):
    """
    Complete curriculum metadata attached to activities.
    This is the SOURCE OF TRUTH for curriculum IDs.
    """
    subject: str = Field(..., description="Subject name")
    unit: CurriculumUnit = Field(..., description="Unit information")
    skill: CurriculumSkill = Field(..., description="Skill information")
    subskill: CurriculumSubskill = Field(..., description="Subskill information")

    class Config:
        json_schema_extra = {
            "example": {
                "subject": "Social Studies",
                "unit": {
                    "id": "SS001",
                    "title": "Classroom Routines and Social Skills",
                    "description": "Classroom Routines and Social Skills"
                },
                "skill": {
                    "id": "SS001-04",
                    "description": "Local Governance"
                },
                "subskill": {
                    "id": "SS001-04-E",
                    "description": "Create simple maps showing locations of local public services"
                }
            }
        }


# ============================================================================
# CURRICULUM QUERY/RESPONSE SCHEMAS
# ============================================================================

class CurriculumItem(BaseModel):
    """Individual curriculum item from BigQuery"""
    subject: str
    grade: Optional[str] = None
    unit_id: str
    unit_title: str
    skill_id: str
    skill_description: str
    subskill_id: str
    subskill_description: str
    difficulty_start: Optional[int] = None
    difficulty_end: Optional[int] = None
    target_difficulty: Optional[int] = None

    def to_curriculum_metadata(self) -> CurriculumMetadata:
        """Convert to CurriculumMetadata schema"""
        return CurriculumMetadata(
            subject=self.subject,
            unit=CurriculumUnit(
                id=self.unit_id,
                title=self.unit_title,
                description=self.unit_title
            ),
            skill=CurriculumSkill(
                id=self.skill_id,
                description=self.skill_description
            ),
            subskill=CurriculumSubskill(
                id=self.subskill_id,
                description=self.subskill_description
            )
        )


# ============================================================================
# HELPER FUNCTIONS
# ============================================================================

def create_curriculum_metadata(
    subject: str,
    unit_id: str,
    unit_title: str,
    skill_id: str,
    skill_description: str,
    subskill_id: str,
    subskill_description: str,
    unit_description: Optional[str] = None
) -> CurriculumMetadata:
    """Helper to create CurriculumMetadata from individual fields"""
    return CurriculumMetadata(
        subject=subject,
        unit=CurriculumUnit(
            id=unit_id,
            title=unit_title,
            description=unit_description or unit_title
        ),
        skill=CurriculumSkill(
            id=skill_id,
            description=skill_description
        ),
        subskill=CurriculumSubskill(
            id=subskill_id,
            description=subskill_description
        )
    )

File: my-tutoring-app/src/types/curriculum.ts
/**
 * Curriculum Type Definitions
 *
 * These types match the backend schema exactly to ensure type safety
 * across the entire application. DO NOT parse or transform these IDs -
 * use them as-is from the curriculum_metadata provided by the backend.
 */

// ============================================================================
// CORE CURRICULUM TYPES (Match backend exactly)
// ============================================================================

export interface CurriculumUnit {
  id: string;          // e.g., "SS001", "COUNT001"
  title: string;       // e.g., "Classroom Routines and Social Skills"
  description?: string;
}

export interface CurriculumSkill {
  id: string;          // e.g., "SS001-04", "COUNT001-01"
  description: string; // e.g., "Local Governance"
}

export interface CurriculumSubskill {
  id: string;          // e.g., "SS001-04-E", "COUNT001-01-A"
  description: string; // e.g., "Create simple maps showing locations of local public services"
}

/**
 * Curriculum metadata attached to activities
 * This is the SOURCE OF TRUTH for curriculum IDs
 */
export interface CurriculumMetadata {
  subject: string;
  unit: CurriculumUnit;
  skill: CurriculumSkill;
  subskill: CurriculumSubskill;
}

// ============================================================================
// ACTIVITY TYPES (Daily/Weekly Planner)
// ============================================================================

export type ActivityType = 'practice' | 'packages' | 'review' | 'tutoring' | 'assessment';
export type ActivityPriority = 'high' | 'medium' | 'low';
export type ActivityPedagogicalType = 'warm_up' | 'core_challenge' | 'practice' | 'cool_down';
export type ActivitySourceType = 'ai_recommendations' | 'bigquery_recommendations' | 'fallback';

export interface ActivitySourceDetails {
  ai_reason?: string;
  priority_rank?: number;
  estimated_time_minutes?: number;
  readiness_status?: string;
  mastery_level?: number;
  reason?: string;
}

export interface CurriculumTransparency {
  subject: string;
  unit: string;
  skill: string;
  subskill: string;
}

/**
 * Activity from Daily Planner or Weekly Planner
 *
 * IMPORTANT:
 * - `id` is for tracking/routing (e.g., "weekly-ACT-Social Studies-0-1")
 * - `curriculum_metadata` contains the ACTUAL curriculum IDs to use for API calls
 * - Never parse or derive curriculum IDs from the activity `id`
 */
export interface DailyActivity {
  id: string;  // Activity instance ID (for routing/tracking only)
  type: ActivityType;
  title: string;
  description: string;
  category: string;
  estimated_time: string;
  points: number;
  priority: ActivityPriority;
  time_slot?: string;
  action?: string;
  endpoint?: string;
  icon_type?: string;
  is_complete?: boolean;

  // THE KEY: Contains actual curriculum IDs - USE THESE for API calls
  curriculum_metadata?: CurriculumMetadata;

  metadata: Record<string, any>;

  // Transparency and recommendation metadata
  source_type?: ActivitySourceType;
  source_details?: ActivitySourceDetails;
  activity_type?: ActivityPedagogicalType;
  reason?: string;
  curriculum_transparency?: CurriculumTransparency;
}

// ============================================================================
// API REQUEST/RESPONSE TYPES
// ============================================================================

/**
 * Problem Generation Request (matches backend ProblemRequest schema)
 *
 * IMPORTANT: Use IDs from curriculum_metadata, not from activity.id
 */
export interface ProblemGenerationRequest {
  subject: string;
  unit_id?: string;      // From curriculum_metadata.unit.id
  skill_id?: string;     // From curriculum_metadata.skill.id
  subskill_id?: string;  // From curriculum_metadata.subskill.id
  difficulty?: number;
  count?: number;
}

/**
 * Topic selection for practice sessions
 * This structure is used by ProblemSet component
 */
export interface TopicSelection {
  subject: string;
  selection: {
    unit?: string;
    skill: string;
    subskill: string;
  };
  unit?: CurriculumUnit;
  skill?: CurriculumSkill;
  subskill?: CurriculumSubskill;
  difficulty_range?: {
    target: number;
  };
  autoStart?: boolean;
  fromCardInterface?: boolean;
}

// ============================================================================
// HELPER FUNCTIONS
// ============================================================================

/**
 * Extract TopicSelection from DailyActivity
 * Uses curriculum_metadata as the source of truth
 */
export function activityToTopicSelection(
  activity: DailyActivity,
  autoStart: boolean = true
): TopicSelection | null {
  if (!activity.curriculum_metadata) {
    console.warn('Activity missing curriculum_metadata:', activity.id);
    return null;
  }

  const { curriculum_metadata } = activity;

  return {
    subject: curriculum_metadata.subject,
    selection: {
      unit: curriculum_metadata.unit.id,
      skill: curriculum_metadata.skill.id,
      subskill: curriculum_metadata.subskill.id,
    },
    unit: curriculum_metadata.unit,
    skill: curriculum_metadata.skill,
    subskill: curriculum_metadata.subskill,
    difficulty_range: {
      target: activity.metadata?.difficulty || 3.0,
    },
    autoStart,
    fromCardInterface: true,
  };
}

/**
 * Extract ProblemGenerationRequest from DailyActivity
 * Uses curriculum_metadata as the source of truth
 */
export function activityToProblemRequest(
  activity: DailyActivity,
  count: number = 5
): ProblemGenerationRequest | null {
  if (!activity.curriculum_metadata) {
    console.warn('Activity missing curriculum_metadata:', activity.id);
    return null;
  }

  const { curriculum_metadata } = activity;

  return {
    subject: curriculum_metadata.subject,
    unit_id: curriculum_metadata.unit.id,
    skill_id: curriculum_metadata.skill.id,
    subskill_id: curriculum_metadata.subskill.id,
    difficulty: activity.metadata?.difficulty,
    count,
  };
}

/**
 * Validate that an activity has proper curriculum metadata
 */
export function hasValidCurriculumMetadata(activity: DailyActivity): boolean {
  return !!(
    activity.curriculum_metadata &&
    activity.curriculum_metadata.subject &&
    activity.curriculum_metadata.unit?.id &&
    activity.curriculum_metadata.skill?.id &&
    activity.curriculum_metadata.subskill?.id
  );
}
